jupyter_client/provisioning.py

```diff
@@ -0,0 +1,407 @@
+"""Kernel Provisioner Classes"""
+
+# Copyright (c) Jupyter Development Team.
+# Distributed under the terms of the Modified BSD License.
+import asyncio
+import os
+import signal
+import sys
+
+from abc import ABCMeta, ABC, abstractmethod
+from entrypoints import EntryPoint, get_group_all, get_single, NoSuchEntryPoint
+from typing import Optional, Dict, List, Any, Tuple
+from traitlets.config import Config, LoggingConfigurable, SingletonConfigurable
+
+from .connect import write_connection_file
+from .launcher import async_launch_kernel
+from .localinterfaces import is_local_ip, local_ips
+
+DEFAULT_PROVISIONER = "LocalProvisioner"
+
+
+class KernelProvisionerMeta(ABCMeta, type(LoggingConfigurable)):
+    pass
+
+
+class KernelProvisionerBase(ABC, LoggingConfigurable, metaclass=KernelProvisionerMeta):
+    """Base class defining methods for KernelProvisioner classes.
+
+       Theses methods model those of the Subprocess Popen class:
+       https://docs.python.org/3/library/subprocess.html#popen-objects
+    """
+    # The kernel specification associated with this provisioner
+    kernel_spec: Any
+    kernel_id: str
+
+    def __init__(self, kernel_id: str, kernel_spec: Any, **kwargs):
+        self.kernel_id = kernel_id
+        self.kernel_spec = kernel_spec
+        super().__init__(**kwargs)
+
+    @abstractmethod
+    async def poll(self) -> [int, None]:
+        """Checks if kernel process is still running.
+
+         If running, None is returned, otherwise the process's integer-valued exit code is returned.
+         """
+        pass
+
+    @abstractmethod
+    async def wait(self) -> [int, None]:
+        """Waits for kernel process to terminate."""
+        pass
+
+    @abstractmethod
+    async def send_signal(self, signum: int) -> None:
+        """Sends signal identified by signum to the kernel process."""
+        pass
+
+    @abstractmethod
+    async def kill(self, restart=False) -> None:
+        """Kills the kernel process.  This is typically accomplished via a SIGKILL signal, which
+        cannot be caught.
+
+        restart is True if this operation precedes a start launch_kernel request.
+        """
+        pass
+
+    @abstractmethod
+    async def terminate(self, restart=False) -> None:
+        """Terminates the kernel process.  This is typically accomplished via a SIGTERM signal, which
+        can be caught, allowing the kernel provisioner to perform possible cleanup of resources.
+
+        restart is True if this operation precedes a start launch_kernel request.
+        """
+        pass
+
+    @abstractmethod
+    async def launch_kernel(self, cmd: List[str], **kwargs: Any) -> Tuple['KernelProvisionerBase', Dict]:
+        """Launch the kernel process returning the class instance and connection info."""
+        pass
+
+    async def cleanup(self, restart=False) -> None:
```

Comment:
Should we force this to be abstract as well?

Reply:
I'm okay with that.  Since the default implementation didn't need this (although the connection file management is a concern that could change that), I didn't want to force its (potentially unnecessary) implementation.  That said, `cleanup()` is something generally expected - with the act of nothing to do being the exception.

Reply:
I'm fine either way. I think forcing implementers to consider what cleanup they need isn't the worst overhead even if it's a `pass` implementation but my opinion there isn't a strong one.